accepted/2020/serializer/WriteableDomAndDynamic.md

```diff
@@ -0,0 +1,577 @@
+// Ask for a number with a JSON number token type.
+// Internally probably stored as a byte[] or JsonElement.
+JsonNumber number = JsonSerializer.Deserialize<JsonNumber>("1.2");
+double dlb = number.GetValue<double>();
+
+// Change to "NaN".
+number.Value = double.Nan;
+string json = JsonSerializer.Serialize(number);
+Debug.Assert(json == "\"NaN\""); // Due to quoted number support, this works.
+
+// However what about round-tripping?
+// Can a JSON string be deserialized into a JsonNumber? If so, "JsonNumber" no longer means a "JSON number".
+// When deserializing a string there is no known CLR type, so the internal value would need be based on a string\byte[].
+// So a "JsonString" node would normally be created, not "JsonNumber" and thus throw here:
+number = JsonSerializer.Deserialize<JsonNumber>(json); // Throws
+
+// Even if the above could be made to work, it wouldn't work with a JsonObject node.
+// A JsonObject would create "JsonString" nodes for "NaN" JSON strings since it has
+// no way of knowing these should map to a JsonNumber (or a CLR number).
+```
+
+Thus the DOM is based on CLR types (not JSON tokens) in edit mode and during serialization. The JSON "value" types (number, string, boolean) are represented as a common `JsonValue` sealed class, and not separate `JsonNumber`, `JsonString` and `JsonBoolean` classes as shown in the example above.
+
+During serialization, the DOM is written using the serializer and uses any custom converters that may be configured. In all but idiocentric scenarios (e.g. a custom converter can do anything), the behavior is natural and JSON is round-tripped as expected.
+
+In the example, a JSON "NaN" string is deserialized into a `JsonValue` and when `GetValue<double>` is called, a `double` is returned, assuming the appropriate `JsonSerializerOptions` are configured for quoted numbers. When the node is serialized, it produces a JSON string (again, if configured). The consumer of the DOM is not necessarily aware that `double` values assigned to the node may be serialized to JSON as either a string or a JSON number, and after deserialization is not aware what the internal state is - it could be a `byte[]`, a `JsonElement` or a `string` depending on implementation and\or the JSON token (string or number).
+
+## Primary API
+These types live in STJ.dll.
+```cs
+namespace System.Text.Json
+// Also consider existing S.T.J.Serialization namespace instead of S.T.J.
+// since it is now coupled to the serializer rather than stand-alone.
+{
+    public abstract class JsonNode
+    {
+        public JsonNode(JsonSerializerOptions options = null);
+
+        public JsonSerializerOptions Options { get; }
+
+        // Return the internal value or convert to T as necessary.
+        public abstract T GetValue<T>();
+
+        // The token type from deserialization; otherwise JsonValueKind.Unspecified.
+        public JsonValueKind ValueKind { get; }
+    }
+
+    public sealed class JsonArray : JsonNode, IList<object>
```

Comment:
`IList<object>`?

Why not `IList<JsonNode>`

Reply:
See https://github.com/dotnet/designs/pull/163/files#diff-34fc37119c1d4491b844207aeed3a9c846c3e2d835fd712e164b1e47b89687c0R358

Reply:
You could get around this by making `JsonNode` have implicit constructors to convert from primitive values to `JsonNode`.

Reply:
Using `object` allows any serializable CLR type to be added to the collection.

A previous prototype did use operators, but for various reasons explained in the doc, we are now trying to avoid implicit (or explicit) operators.

Reply:
The doc does mention some reasons why implicit operators are not used (not supported in all languages, consistency for types that we don't know, not needed when using dynamic).

Reply:
> You could get around this by making `JsonNode` have implicit constructors to convert from primitive values to `JsonNode`.

This is how the initial prototype solved the issue, but kept `JsonArray` and `IList<JsonNode>`:
https://github.com/ahsonkhan/corefx/blob/5bc2806f33090e78b38fafe4d5b46d5a0a4c1f08/src/System.Text.Json/ref/System.Text.Json.cs#L240-L259

> not supported in all languages

Does that mean we shouldn't have any implicit operators in the framework?

> consistency for types that we don't know

What do you mean by this?

Reply:
> Does that mean we shouldn't have any implicit operators in the framework?

For serialization, yes IMHO. Partly because it results in an inconsistent programming model (see below; also applies to explicit operators) but also because an implicit operator should not throw and it should be a simple, lossless, fast conversion which doesn't hold for serialization.

>> consistency for types that we don't know

> What do you mean by this?

We can only add operators to known types such as `string` etc. We can't add operators to custom types (e.g. a third party `Currency` struct or a `DateTimeOffset` custom converter). So by not supporting operators we level the "known" types along with the "unknown" types and thus have a consistent programming model across both.